File: android/src/main/java/com/audiocontext/nodes/StereoPannerNode.kt
package com.audiocontext.nodes

import android.os.Build
import androidx.annotation.RequiresApi
import com.audiocontext.context.BaseAudioContext
import com.audiocontext.nodes.audionode.AudioNode
import com.audiocontext.nodes.audionode.ChannelCountMode
import com.audiocontext.parameters.AudioParam
import com.audiocontext.parameters.PlaybackParameters
import com.audiocontext.utils.Constants
import kotlin.math.*

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

class StereoPannerNode(context: BaseAudioContext): AudioNode(context) {
  override val numberOfInputs: Int = 1
  override val numberOfOutputs: Int = 1
  override var channelCount: Int = 2
  override val channelCountMode: ChannelCountMode = ChannelCountMode.CLAMPED_MAX

  private val pan: AudioParam = AudioParam(context,0.0, Constants.MAX_PAN, -Constants.MAX_PAN)
    get() = field

  @RequiresApi(Build.VERSION_CODES.N)
  override fun process(playbackParameters: PlaybackParameters) {
    mixBuffers(playbackParameters)

    val pan = pan.getValueAtTime(context.getCurrentTime())
    val x = (pan + 1) / 2

    val gainL = cos(x * PI / 2)
    val gainR = sin(x * PI / 2)

    playbackParameters.leftPan *= gainL
    playbackParameters.rightPan *= gainR

    super.process(playbackParameters)
  }
}

File: packages/react-native-audio-api/android/src/main/cpp/StereoPannerNode/StereoPannerNode.cpp
#include "StereoPannerNode.h"
#include "AudioContext.h"

// https://webaudio.github.io/web-audio-api/#stereopanner-algorithm

namespace audioapi {

StereoPannerNode::StereoPannerNode(AudioContext *context) : AudioNode(context) {
  channelCountMode_ = ChannelCountMode::CLAMPED_MAX;
  panParam_ = std::make_shared<AudioParam>(context, 0.0, -MAX_PAN, MAX_PAN);
}

std::shared_ptr<AudioParam> StereoPannerNode::getPanParam() const {
  return panParam_;
}

void StereoPannerNode::processAudio() {
    //channelCount = 2
    AudioNode::processAudio();

    for (int i = 0; i < context_->getBufferSize(); i++) {
        auto pan = panParam_->getValueAtTime(context_->getCurrentTime());
        auto x = (pan <= 0 ? pan + 1 : pan) * M_PI / 2;

        auto gainL = static_cast<float>(cos(x));
        auto gainR = static_cast<float>(sin(x));

        auto inputL = inputBuffer_[i * 2];
        auto inputR = inputBuffer_[i * 2 + 1];

    if (pan <= 0) {
      outputBuffer_[i * 2] = inputL + inputR * gainL;
      outputBuffer_[i * 2 + 1] = inputR * gainR;
    } else {
        outputBuffer_[i * 2] = inputL * gainL;
        outputBuffer_[i * 2 + 1] = inputR + inputL * gainR;
    }
  }
}

} // namespace audioapi
